### PR TITLE
cilium, bigtcp: Make probing for GRO/GSO max size more graceful

### PR DESCRIPTION
**Pull request by borkmann**
See commits.

### CODE REVIEW

**Comment by borkmann** [comment]
Cc @haiyuewa

**Comment by haiyuewa** [comment]
@borkmann Cool, small code, big improvement:

IPv4 BIG TCP, IPv6 BIG TCP disabled:
kubectl exec netperf-client -- netperf  -t TCP_RR -H ${NETPERF_SERVER_IPV6} -- -r80000:80000 -O MIN_LATENCY,P90_LATENCY,P99_LATENCY,THROUGHPUT
MIGRATED TCP REQUEST/RESPONSE TEST from ::0 (::) port 0 AF_INET6 to 2013:2013:2013:1::acd2 () port 0 AF_INET6 : first burst 0
Minimum      90th         99th         Throughput
Latency      Percentile   Percentile
Microseconds Latency      Latency
             Microseconds Microseconds
86           273          706          **4782.05**


kubectl exec netperf-client -- netperf  -t TCP_RR -H ${NETPERF_SERVER_IPV4} -- -r80000:80000 -O MIN_LATENCY,P90_LATENCY,P99_LATENCY,THROUGHPUT
MIGRATED TCP REQUEST/RESPONSE TEST from 0.0.0.0 (0.0.0.0) port 0 AF_INET to 13.13.1.97 (13.13.1) port 0 AF_INET : first burst 0
Minimum      90th         99th         Throughput
Latency      Percentile   Percentile
Microseconds Latency      Latency
             Microseconds Microseconds
86           319          772          **4441.95**

============================================

IPv4 BIG TCP, IPv6 BIG TCP enabled:

kubectl -n kube-system logs ds/cilium 2>&1 | grep "big-tcp"
level=info msg="  --enable-ipv4-big-tcp='true'" subsys=daemon
level=info msg="  --enable-ipv6-big-tcp='true'" subsys=daemon
level=info msg="Setting up BIG TCP" subsys=big-tcp
level=info msg="Lowering GRO/GSO max size from 196608 to 131072" device=enp0 subsys=big-tcp
level=info msg="Setting IPv6 gso_max_size to 131072 and gro_max_size to 131072" device=enp0 subsys=big-tcp
level=info msg="Setting IPv4 gso_max_size to 131072 and gro_max_size to 131072" device=enp0 subsys=big-tcp

kubectl exec netperf-server -- ip -d -j link show dev eth0 | jq -c '.[0].gso_max_size'
131072

kubectl exec netperf-client -- ip -d -j link show dev eth0 | jq -c '.[0].gso_max_size'
131072

ip -d -j link show dev enp0 | jq -c '.[0].gso_max_size'
131072

ip -d -j link show dev enp0 | jq -c '.[0].gso_ipv4_max_size'
131072

kubectl exec netperf-client -- netperf  -t TCP_RR -H ${NETPERF_SERVER_IPV6} -- -r80000:80000 -O MIN_LATENCY,P90_LATENCY,P99_LATENCY,THROUGHPUT
MIGRATED TCP REQUEST/RESPONSE TEST from ::0 (::) port 0 AF_INET6 to 2013:2013:2013:1::a041 () port 0 AF_INET6 : first burst 0
Minimum      90th         99th         Throughput
Latency      Percentile   Percentile
Microseconds Latency      Latency
             Microseconds Microseconds
75           158          651          **7383.83**


kubectl exec netperf-client -- netperf  -t TCP_RR -H ${NETPERF_SERVER_IPV4} -- -r80000:80000 -O MIN_LATENCY,P90_LATENCY,P99_LATENCY,THROUGHPUT
MIGRATED TCP REQUEST/RESPONSE TEST from 0.0.0.0 (0.0.0.0) port 0 AF_INET to 13.13.1.177 (13.13.1) port 0 AF_INET : first burst 0
Minimum      90th         99th         Throughput
Latency      Percentile   Percentile
Microseconds Latency      Latency
             Microseconds Microseconds
74           146          453          **7791.12**

**Comment by borkmann** [comment]
Awesome, thanks @haiyuewa!

**Comment by borkmann** [comment]
(CI all green, no code change. Adding a small doc update to list ice.)

File: pkg/datapath/linux/bigtcp/bigtcp.go [hunk 1]
```diff
@@ -137,6 +138,22 @@ func haveIPv6MaxSize() bool {
...
 }
 
+func probeTSOMaxSize() int {
+	maxSize := math.IntMin(bigTCPGSOMaxSize, bigTCPGROMaxSize)
+	for _, device := range option.Config.GetDevices() {
+		link, err := netlink.LinkByName(device)
+		if err == nil {
+			tso := link.Attrs().TSOMaxSize
+			tsoMax := int(tso)
+			if tsoMax > defaultGSOMaxSize && tsoMax < maxSize {
```
Comment (aspsk) [inline pkg/datapath/linux/bigtcp/bigtcp.go:148]: Could it happen that `link.Attrs().TSOMaxSize` is less than 65536? If yes, shouldn't we set `maxSize` to the lower value? If not, then the `tsoMax > defaultGSOMaxSize` check is not required.

Comment (borkmann) [reply pkg/datapath/linux/bigtcp/bigtcp.go:148]: If it is less than that, we get a failure later and won't touch it further.

**Review by aspsk — APPROVED** [review]

**Pull request merged**